airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/global_substream_cursor.py

```diff
@@ -4,13 +4,43 @@
 
 import threading
 import time
-from typing import Any, Iterable, Mapping, Optional, Union
+from typing import Any, Iterable, Mapping, Optional, TypeVar, Union
 
 from airbyte_cdk.sources.declarative.incremental.datetime_based_cursor import DatetimeBasedCursor
 from airbyte_cdk.sources.declarative.incremental.declarative_cursor import DeclarativeCursor
 from airbyte_cdk.sources.declarative.partition_routers.partition_router import PartitionRouter
 from airbyte_cdk.sources.types import Record, StreamSlice, StreamState
 
+T = TypeVar("T")
+
+
+def iterate_with_last_flag(generator: Iterable[T]) -> Iterable[tuple[Optional[T], bool]]:
```

Comment:
would it be valuable to add a few tests for this new method to `test_parent_stream_state.py`? This might not be necessary since its invoked from some of the other methods being tested, but if there are any specific edge cases, it might be easier to unit test it at the lowest level

Reply:
I don't think we need dedicated unit tests for `iterate_with_flag` since it's a simple utility function and its behavior is clearly documented—it just adds a flag for the last element and returns `(None, True)` for an empty generator. Instead, I've focused on edge cases where it's used, adding tests like `test_incremental_parent_state_no_slices` and `test_incremental_parent_state_no_records` in `test_parent_stream_state.py` to ensure it handles empty generators correctly in context. Let me know if you still feel more specific tests are needed.

Reply:
fair enough we can skip